FastAPI background tasks broken after refactoring in 1.7.0

One of the refactorings in `1.7.0`, probably #452, broke FastAPI background tasks. 

stack trace:
```
AssertionError: 
[ERROR] Application callable raised an exception
  File "REDACTED/.venv/lib/python3.12/site-packages/granian/_futures.py", line 11, in future_watcher
    await inner(watcher.scope, watcher.proto)
  File "REDACTED/.venv/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/applications.py", line 113, in __call__
    await self.middleware_stack(scope, receive, send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 187, in __call__
    raise exc
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 165, in __call__
    await self.app(scope, receive, _send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    raise exc
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    await app(scope, receive, sender)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/routing.py", line 715, in __call__
    await self.middleware_stack(scope, receive, send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/routing.py", line 735, in app
    await route.handle(scope, receive, send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/routing.py", line 288, in handle
    await self.app(scope, receive, send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/routing.py", line 76, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    raise exc
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    await app(scope, receive, sender)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/routing.py", line 74, in app
    await response(scope, receive, send)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/responses.py", line 158, in __call__
    await self.background()
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/background.py", line 41, in __call__
    await task()
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/background.py", line 28, in __call__
    await run_in_threadpool(self.func, *self.args, **self.kwargs)
  File "REDACTED/.venv/lib/python3.12/site-packages/starlette/concurrency.py", line 39, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "REDACTED/.venv/lib/python3.12/site-packages/anyio/to_thread.py", line 56, in run_sync
    return await get_async_backend().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "REDACTED/.venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 2469, in run_sync_in_worker_thread
    async with limiter or cls.current_default_thread_limiter():
  File "REDACTED/.venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 1982, in __aenter__
    await self.acquire()
  File "REDACTED/.venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 2036, in acquire
    return await self.acquire_on_behalf_of(current_task())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "REDACTED/.venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 2039, in acquire_on_behalf_of
    await AsyncIOBackend.checkpoint_if_cancelled()
  File "REDACTED/.venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 2372, in checkpoint_if_cancelled
    cancel_scope = _task_states[task].cancel_scope
                   ~~~~~~~~~~~~^^^^^^
  File "REDACTED/.venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 686, in __getitem__
    assert isinstance(key, asyncio.Task)
```

The reason is a [wrapper](https://github.com/emmett-framework/granian/pull/452/files#diff-571aaa34f48968993dca3f59fcc89aff6e892451c088006d134ab11c0acf0cc2R223) that has been introduced and a very restrictive `assert`.

## Repro

Create the following app:

```
from fastapi import FastAPI, BackgroundTasks
from fastapi.responses import JSONResponse
import logging

logger = logging.getLogger(__name__)

app = FastAPI()

@app.get("/hello")
async def hello(background_tasks: BackgroundTasks) -> JSONResponse:
    def background_task():
        logger.info("run background task")

    background_tasks.add_task(background_task)
    
    return JSONResponse(status_code=200, content={"message": "Hello world"})
```

run it:

`granian --interface asgi main:app`

query:

`curl http://127.0.0.1:8000/hello`

1.6.4: :heavy_check_mark: 
1.7.0: :stop_sign:

--- a/README.md
+++ b/README.md
@@ -101,6 +101,9 @@ Options:
   --loop [auto|asyncio|rloop|uvloop]
                                   Event loop implementation  [env var:
                                   GRANIAN_LOOP; default: (auto)]
+  --task-impl [auto|rust|asyncio]
+                                  Async task implementation to use  [env var:
+                                  GRANIAN_TASK_IMPL; default: (auto)]
   --backlog INTEGER RANGE         Maximum number of connections to hold in
                                   backlog (globally)  [env var:
                                   GRANIAN_BACKLOG; default: 1024; x>=128]

--- a/granian/_futures.py
+++ b/granian/_futures.py
@@ -25,10 +25,17 @@ class _CBScheduler(_BaseCBScheduler):
         self._schedule_fn = _cbsched_schedule(loop, ctx, self._run, cb)
 
 
-def _new_cbscheduler(loop, cb):
-    return _CBScheduler(
-        loop, contextvars.copy_context(), cb, partial(_aio_taskenter, loop), partial(_aio_taskleave, loop)
-    )
+class _CBSchedulerAIO(_BaseCBScheduler):
+    __slots__ = []
+
+    def __init__(self, loop, ctx, cb, aio_tenter, aio_texit):
+        super().__init__()
+        self._schedule_fn = _cbsched_schedule(loop, ctx, loop.create_task, cb)
+
+
+def _new_cbscheduler(loop, cb, impl_asyncio=False):
+    _cls = _CBSchedulerAIO if impl_asyncio else _CBScheduler
+    return _cls(loop, contextvars.copy_context(), cb, partial(_aio_taskenter, loop), partial(_aio_taskleave, loop))
 
 
 def _cbsched_schedule(loop, ctx, run, cb):

--- a/granian/_imports.py
+++ b/granian/_imports.py
@@ -1,4 +1,9 @@
 try:
+    import anyio
+except ImportError:
+    anyio = None
+
+try:
     import setproctitle
 except ImportError:
     setproctitle = None

--- a/granian/cli.py
+++ b/granian/cli.py
@@ -5,7 +5,7 @@ from typing import Any, Callable, List, Optional, Type, TypeVar, Union
 
 import click
 
-from .constants import HTTPModes, Interfaces, Loops, ThreadModes
+from .constants import HTTPModes, Interfaces, Loops, TaskImpl, ThreadModes
 from .errors import FatalError
 from .http import HTTP1Settings, HTTP2Settings
 from .log import LogLevels
@@ -77,6 +77,12 @@ def option(*param_decls: str, cls: Optional[Type[click.Option]] = None, **attrs:
     help='Threading mode to use',
 )
 @option('--loop', type=EnumType(Loops), default=Loops.auto, help='Event loop implementation')
+@option(
+    '--task-impl',
+    type=EnumType(TaskImpl),
+    default=TaskImpl.auto,
+    help='Async task implementation to use',
+)
 @option(
     '--backlog',
     type=click.IntRange(128),
@@ -261,6 +267,7 @@ def cli(
     blocking_threads: Optional[int],
     threading_mode: ThreadModes,
     loop: Loops,
+    task_impl: TaskImpl,
     backlog: int,
     backpressure: Optional[int],
     http1_buffer_size: int,
@@ -316,6 +323,7 @@ def cli(
         blocking_threads=blocking_threads,
         threading_mode=threading_mode,
         loop=loop,
+        task_impl=task_impl,
         http=http,
         websockets=websockets,
         backlog=backlog,

--- a/granian/constants.py
+++ b/granian/constants.py
@@ -29,3 +29,9 @@ class Loops(StrEnum):
     asyncio = 'asyncio'
     rloop = 'rloop'
     uvloop = 'uvloop'
+
+
+class TaskImpl(StrEnum):
+    auto = 'auto'
+    rust = 'rust'
+    asyncio = 'asyncio'

--- a/granian/server.py
+++ b/granian/server.py
@@ -14,11 +14,11 @@ from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Type
 
 from ._futures import _future_watcher_wrapper, _new_cbscheduler
 from ._granian import ASGIWorker, RSGIWorker, WSGIWorker
-from ._imports import setproctitle, watchfiles
+from ._imports import anyio, setproctitle, watchfiles
 from ._internal import load_target
 from ._signals import set_main_signals
 from .asgi import LifespanProtocol, _callback_wrapper as _asgi_call_wrap
-from .constants import HTTPModes, Interfaces, Loops, ThreadModes
+from .constants import HTTPModes, Interfaces, Loops, TaskImpl, ThreadModes
 from .errors import ConfigurationError, PidFileError
 from .http import HTTP1Settings, HTTP2Settings
 from .log import DEFAULT_ACCESSLOG_FMT, LogLevels, configure_logging, logger
@@ -81,6 +81,7 @@ class Granian:
         blocking_threads: Optional[int] = None,
         threading_mode: ThreadModes = ThreadModes.workers,
         loop: Loops = Loops.auto,
+        task_impl: TaskImpl = TaskImpl.auto,
         http: HTTPModes = HTTPModes.auto,
         websockets: bool = True,
         backlog: int = 1024,
@@ -118,6 +119,7 @@ class Granian:
         self.threads = max(1, threads)
         self.threading_mode = threading_mode
         self.loop = loop
+        self.task_impl = task_impl
         self.http = http
         self.websockets = websockets
         self.backlog = max(128, backlog)
@@ -188,6 +190,7 @@ class Granian:
         blocking_threads: int,
         backpressure: int,
         threading_mode: ThreadModes,
+        task_impl: TaskImpl,
         http_mode: HTTPModes,
         http1_settings: Optional[HTTP1Settings],
         http2_settings: Optional[HTTP2Settings],
@@ -225,7 +228,9 @@ class Granian:
             *ssl_ctx,
         )
         serve = getattr(worker, {ThreadModes.runtime: 'serve_rth', ThreadModes.workers: 'serve_wth'}[threading_mode])
-        scheduler = _new_cbscheduler(loop, _future_watcher_wrapper(wcallback))
+        scheduler = _new_cbscheduler(
+            loop, _future_watcher_wrapper(wcallback), impl_asyncio=task_impl == TaskImpl.asyncio
+        )
         serve(scheduler, loop, shutdown_event)
 
     @staticmethod
@@ -239,6 +244,7 @@ class Granian:
         blocking_threads: int,
         backpressure: int,
         threading_mode: ThreadModes,
+        task_impl: TaskImpl,
         http_mode: HTTPModes,
         http1_settings: Optional[HTTP1Settings],
         http2_settings: Optional[HTTP2Settings],
@@ -283,7 +289,9 @@ class Granian:
             *ssl_ctx,
         )
         serve = getattr(worker, {ThreadModes.runtime: 'serve_rth', ThreadModes.workers: 'serve_wth'}[threading_mode])
-        scheduler = _new_cbscheduler(loop, _future_watcher_wrapper(wcallback))
+        scheduler = _new_cbscheduler(
+            loop, _future_watcher_wrapper(wcallback), impl_asyncio=task_impl == TaskImpl.asyncio
+        )
         serve(scheduler, loop, shutdown_event)
         loop.run_until_complete(lifespan_handler.shutdown())
 
@@ -298,6 +306,7 @@ class Granian:
         blocking_threads: int,
         backpressure: int,
         threading_mode: ThreadModes,
+        task_impl: TaskImpl,
         http_mode: HTTPModes,
         http1_settings: Optional[HTTP1Settings],
         http2_settings: Optional[HTTP2Settings],
@@ -343,7 +352,9 @@ class Granian:
             *ssl_ctx,
         )
         serve = getattr(worker, {ThreadModes.runtime: 'serve_rth', ThreadModes.workers: 'serve_wth'}[threading_mode])
-        scheduler = _new_cbscheduler(loop, _future_watcher_wrapper(callback))
+        scheduler = _new_cbscheduler(
+            loop, _future_watcher_wrapper(callback), impl_asyncio=task_impl == TaskImpl.asyncio
+        )
         serve(scheduler, loop, shutdown_event)
         callback_del(loop)
 
@@ -358,6 +369,7 @@ class Granian:
         blocking_threads: int,
         backpressure: int,
         threading_mode: ThreadModes,
+        task_impl: TaskImpl,
         http_mode: HTTPModes,
         http1_settings: Optional[HTTP1Settings],
         http2_settings: Optional[HTTP2Settings],
@@ -385,7 +397,9 @@ class Granian:
             worker_id, sfd, threads, blocking_threads, backpressure, http_mode, http1_settings, http2_settings, *ssl_ctx
         )
         serve = getattr(worker, {ThreadModes.runtime: 'serve_rth', ThreadModes.workers: 'serve_wth'}[threading_mode])
-        scheduler = _new_cbscheduler(loop, _wsgi_call_wrap(callback, scope_opts, log_access_fmt))
+        scheduler = _new_cbscheduler(
+            loop, _wsgi_call_wrap(callback, scope_opts, log_access_fmt), impl_asyncio=task_impl == TaskImpl.asyncio
+        )
         serve(scheduler, loop, shutdown_event)
         shutdown_event.qs.wait()
 
@@ -416,6 +430,7 @@ class Granian:
                 self.blocking_threads,
                 self.backpressure,
                 self.threading_mode,
+                self.task_impl,
                 self.http,
                 self.http1_settings,
                 self.http2_settings,
@@ -712,6 +727,9 @@ class Granian:
             if self.workers_lifetime < 60:
                 logger.error('Workers lifetime cannot be less than 60 seconds')
                 raise ConfigurationError('workers_lifetime')
+
+        if self.task_impl == TaskImpl.auto:
+            self.task_impl = TaskImpl.asyncio if anyio is not None else TaskImpl.rust
 
         serve_method = self._serve_with_reloader if self.reload_on_changes else self._serve
         serve_method(spawn_target, target_loader)

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -98,7 +98,7 @@ extend-ignore = [
     'S110', # except pass is fine
 ]
 flake8-quotes = { inline-quotes = 'single', multiline-quotes = 'double' }
-mccabe = { max-complexity = 13 }
+mccabe = { max-complexity = 14 }
 
 [tool.ruff.format]
 quote-style = 'single'

